File: node/src/providers/gemini.ts
import { GoogleGenAI as PostHogGoogleGenAI } from '@posthog/ai';
import { PostHog } from 'posthog-node';
import { BaseProvider, Tool } from './base.js';
import { GEMINI_MODEL, DEFAULT_POSTHOG_DISTINCT_ID } from './constants.js';

export class GeminiProvider extends BaseProvider {
  private client: any;
  private history: any[] = [];
  private config: any;

  constructor(posthogClient: PostHog, aiSessionId: string | null = null) {
    super(posthogClient, aiSessionId);
    this.client = new PostHogGoogleGenAI({
      apiKey: process.env.GEMINI_API_KEY!,
      // vertexai: true,
      // project: "project-id",
      // location: "us-central1",
      posthog: posthogClient
    });
    
    this.config = {
      tools: this.tools
    };
  }

  protected getToolDefinitions(): Tool[] {
    const weatherFunction = {
      name: 'get_current_weather',
      description: 'Gets the current weather for a given location.',
      parameters: {
        type: 'object',
        properties: {
          latitude: {
            type: 'number',
            description: 'The latitude of the location (e.g., 37.7749 for San Francisco)',
          },
          longitude: {
            type: 'number',
            description: 'The longitude of the location (e.g., -122.4194 for San Francisco)',
          },
          location_name: {
            type: 'string',
            description: 'A human-readable name for the location (e.g., \'San Francisco, CA\' or \'Dublin, Ireland\')',
          },
        },
        required: ['latitude', 'longitude', 'location_name'],
      },
    };

    const jokeFunction = {
      name: 'tell_joke',
      description: 'Tell a joke with a question-style setup and an answer punchline',
      parameters: {
        type: 'object',
        properties: {
          setup: {
            type: 'string',
            description: 'The setup of the joke, usually in question form',
          },
          punchline: {
            type: 'string',
            description: 'The punchline or answer to the joke',
          },
        },
        required: ['setup', 'punchline'],
      },
    };

    return [{
      functionDeclarations: [weatherFunction, jokeFunction]
    }] as any;
  }

  getName(): string {
    return 'Google Gemini';
  }

  resetConversation(): void {
    this.history = [];
  }

  async chat(userInput: string, base64Image?: string): Promise<string> {
    // Build content parts for this message
    let parts: any[];
    
    if (base64Image) {
      // Use native Gemini format for images
      parts = [
        { text: userInput },
        {
          inlineData: {
            mimeType: 'image/png',
            data: base64Image
          }
        }
      ];
    } else {
      // Text-only content
      parts = [{ text: userInput }];
    }
    
    // Add user message to history
    this.history.push({
      role: 'user',
      parts: parts
    });

    const requestParams = {
      model: GEMINI_MODEL,
      posthogDistinctId: process.env.POSTHOG_DISTINCT_ID || DEFAULT_POSTHOG_DISTINCT_ID,
      posthogProperties: {
        $ai_span_name: "gemini_generate_content",
        ...this.getPostHogProperties(),
      },
      contents: this.history,
      config: this.config
    };

    const message = await this.client.models.generateContent(requestParams);
    this.debugApiCall("Google Gemini", requestParams, message);

    const displayParts: string[] = [];
    const modelParts: any[] = [];
    const toolResults: string[] = [];

    if (message.candidates) {
      for (const candidate of message.candidates) {
        if (candidate.content) {
          for (const part of candidate.content.parts) {
            if (part.functionCall) {
              const functionCall = part.functionCall;
              modelParts.push({ functionCall });
              
              if (functionCall.name === 'get_current_weather') {
                const latitude = functionCall.args?.latitude || 0.0;
                const longitude = functionCall.args?.longitude || 0.0;
                const locationName = functionCall.args?.location_name;
                const weatherResult = await this.getWeather(latitude, longitude, locationName);
                const toolResultText = this.formatToolResult('get_weather', weatherResult);
                toolResults.push(toolResultText);
                displayParts.push(toolResultText);
              } else if (functionCall.name === 'tell_joke') {
                const setup = functionCall.args?.setup || '';
                const punchline = functionCall.args?.punchline || '';
                const jokeResult = this.tellJoke(setup, punchline);
                const toolResultText = this.formatToolResult('tell_joke', jokeResult);
                toolResults.push(toolResultText);
                displayParts.push(toolResultText);
              }
            } else if (part.text) {
              modelParts.push({ text: part.text });
              displayParts.push(part.text);
            }
          }
        }
      }
    }

    // Add model response to history
    if (modelParts.length > 0) {
      this.history.push({
        role: 'model',
        parts: modelParts
      });
    }
    
    for (const toolResult of toolResults) {
      this.history.push({
        role: 'model',
        parts: [{ text: `Tool result: ${toolResult}` }]
      });
    }

    return displayParts.length > 0 ? displayParts.join('\n\n') : (message.text || 'No response received');
  }
}

File: python/providers/gemini.py
import os
from posthog.ai.gemini import Client
from posthog import Posthog
from google.genai import types
from .base import BaseProvider

class GeminiProvider(BaseProvider):
    def __init__(self, posthog_client: Posthog):
        super().__init__(posthog_client)
        self.client = Client(
            api_key=os.getenv("GEMINI_API_KEY"),
            # vertexai=True,
            # project="project-id",
            # location="us-central1",
            posthog_client=posthog_client
        )
        # Store conversation history in Gemini's native format
        self.history = []
        
        # Configure tools using proper Google GenAI types
        weather_function = self.get_tool_definitions()[0]
        self.tools = types.Tool(function_declarations=[weather_function])
        self.config = types.GenerateContentConfig(tools=[self.tools])
    
    def get_name(self):
        return "Google Gemini"
    
    def get_tool_definitions(self):
        """Return tool definitions in Gemini format"""
        return [
            {
                "name": "get_current_weather",
                "description": "Gets the current weather for a given location.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city name, e.g. San Francisco",
                        },
                    },
                    "required": ["location"],
                },
            }
        ]
    
    def reset_conversation(self):
        """Reset the conversation history"""
        self.history = []
        self.messages = []
    
    def chat(self, user_input: str, base64_image: str = None) -> str:
        """Send a message to Gemini and get response"""
        # Build content parts for this message
        if base64_image:
            # Use native Gemini format for images
            parts = [
                {"text": user_input},
                {
                    "inline_data": {
                        "mime_type": "image/png",
                        "data": base64_image
                    }
                }
            ]
        else:
            # Text-only content
            parts = [{"text": user_input}]
        
        # Add user message to history
        self.history.append({
            "role": "user",
            "parts": parts
        })

        # Prepare API request parameters
        request_params = {
            "model": "gemini-2.5-flash",
            "posthog_distinct_id": os.getenv("POSTHOG_DISTINCT_ID", "user-hog"),
            "contents": self.history,
            "config": self.config
        }

        # Send all messages in conversation history
        message = self.client.models.generate_content(**request_params)

        # Debug: Log the API call (request + response)
        self._debug_api_call("Google Gemini", request_params, message)

        # Collect response parts for display
        display_parts = []
        model_parts = []
        tool_results = []
        
        # Check if Gemini wants to use tools and collect all response parts
        if hasattr(message, 'candidates') and message.candidates:
            for candidate in message.candidates:
                if hasattr(candidate, 'content') and candidate.content:
                    for part in candidate.content.parts:
                        if hasattr(part, 'function_call') and part.function_call:
                            function_call = part.function_call
                            model_parts.append({"function_call": function_call})
                            
                            if function_call.name == "get_current_weather":
                                location = function_call.args.get("location", "unknown")
                                weather_result = self.get_weather(location)
                                tool_result_text = self.format_tool_result("get_weather", weather_result)
                                tool_results.append(tool_result_text)
                                display_parts.append(tool_result_text)
                        elif hasattr(part, 'text'):
                            model_parts.append({"text": part.text})
                            display_parts.append(part.text)
        
        # Add model response to history
        if model_parts:
            self.history.append({
                "role": "model",
                "parts": model_parts
            })
        
        # Add tool results to history
        for tool_result in tool_results:
            self.history.append({
                "role": "model",
                "parts": [{"text": f"Tool result: {tool_result}"}]
            })
        
        return "\n\n".join(display_parts) if display_parts else (message.text if hasattr(message, 'text') else "No response received")